### ISSUE TITLE
usleep Statments in Cache Loader causing Pathological Page Load Times

### ISSUE THREAD

**Issue by astorm** [magento/magento2#22824]
### Preconditions (*)
1. Magento 2.3.1

### Steps to reproduce (*)
1. Clear Cache
2. Load Page

### Expected result (*)
Page loads in less than 30 - 45 seconds

### Actual result (*)
Page loads in 30 - 45 seconds under unknown circumstances

If this is possible, I'd like to nominate a bit of code for removal/refactoring.  It looks like it was introduced in Magento 2.3.1

https://github.com/magento/magento2/commit/acdc770ac25c6f9d6c621fd7ba44add38bb89b05#diff-cbe30020f802e5ab5b7cf6399e91e178

And currently lives here: https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php#L72

I've had reports from users/customers (who will hopefully be able to comment below) of one of my extensions ([Commerce Bug](https://store.pulsestorm.net/products/commerce-bug-3)) that they're seeing uncached Magento 2 page loads taking between 30 and 45 seconds, with the majority of this being spent [in this call to `usleep`](https://github.com/magento/magento2/blob/2.3.1/app/code/Magento/Config/App/Config/Type/System.php#L239).  The delay is inconsistent and seems dependent on resource contention in the host system (i.e. it may only show up when/if file loading is slower/faster than normal, a certain number of simultaneous requests is present, etc)

I'm reasonably certain my extension isn't to blame -- it's just that my extension makes an early call to `Magento\Developer\Helper\Data::isDevLoaded` in an event observer, and is the first thing to trigger this cache being loaded. As mentioned the problem is inconsistent, so I can't be fully certain what is or isn't to blame.  

It also seems like at least one other extension vendor has this problem.

https://twitter.com/jissereitsma/status/1126746140534165504

Beyond this being a specific problem that real users have run into -- I don't think this is the right solution for the problem, (the problem as listed in the commit comment)

> MC-6275: Conflict of simultaneous write in Redis cache

1. The pattern in use here may be OK in a language with asynchronous callbacks, but PHP's execution model is single threaded and synchronous.  This means the call to `usleep` blocks _everything in the request_.  This is what's causing the 30 - 45 seconds uncached load times for users. In most organizations I've worked for and projects I've worked on a sleep statement of any-kind in user facing PHP code would have a high bar to make it into the repo -- while MC-6275 is an important issue to solve, I'm not sure this is the right solution for this problem

2. If this code is meant to solve simultaneous Redis writes -- then it belong _in the redis cache implementation_.  Having this code up here means that every cache implementation (Redis, Filesystem, etc) is effected.  

3. The recent change to move this from a .1 second sleep to a .02 second sleep are a welcome improvement, but this seems like a work around that's destined to cause future problem for users and customers.

**Comment by amenk** [comment]
Unfortunately I cannot reproduce the problem anymore on my system, but I have the blackfire dump and I can share it privately, because I am not sure if it contains any protected information. Contact: a dot menk at imi dot de

**Comment by IvanChepurnyi** [comment]
Seems this is is the same issue I was working on based on @amenk tweet. Here is a pull request #22829  that fixes this issue

**Comment by sdzhepa** [comment]
Hello @astorm 

Thank you for your report!

I investigated this issue and want to admit it is very hard to find manual scenario/steps to reproduce and confirm the issue.
But I see there is proposed Pull Request #22829 that should fix the problem.

Feel free to collaborate and assist @IvanChepurnyi (Author of fix) and @kandy (expert on Magento performance track) under #22829 discussion. It will really help to review code and deliver the fix as soon as possible.

**Comment by magento-engcom-team** [comment]
@sdzhepa Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess
- [ ] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
@sdzhepa Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @sdzhepa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-99708` were created

**Issue Available**: @sdzhepa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by matei** [comment]
We are also seeing this issue on Magento Commerce 2.3.2 in a not-yet-production environment. Several traces in NewRelic show 5-6 seconds spent in lockedLoadData, sleeping. 

I'm not sure if this is due to the fact that the lock implementation is non-atomic and we end up in some near-deadlocking scenarios, or other issues, but I can confirm the issue is happening and doesn't look related to any extension

**Comment by IvanChepurnyi** [comment]
@matei see PR #22829, it is fixing this issue, but not yet merged into develop branch

**Comment by ilnytskyi** [comment]
Redirected from #22504 
Seems to be the same issue https://github.com/magento/magento2/issues/22504#issuecomment-573582674

**Comment by vzabaznov** [comment]
Hi @astorm and community, thank you for reporting this.

In the process of testing our locking mechanism we found couple areas to improve. We believe this will resolve current performance issues of locking functionality. Internal tickets MC-32426 & MC-32427 were created. Community can expect to have it in the nearest 2.4 release. Also, some fixes to resolve race conditions and lock related fixes were already delivered, community can expect to have them in 2.3.5 and 2.4.

Alternative solution for non-blocking cache revalidation provided by @IvanChepurnyi was put on public architecture proposal process - https://github.com/magento/architecture/pull/369

Since all related tickets already created we will close this.

**Comment by kolaente** [comment]
It looks like this is happening again. This time it seems related to customer sessions instead of caching. Not directly in Magento but [here](https://github.com/colinmollenhour/php-redis-session-abstract/blob/master/src/Cm/RedisSession/Handler.php#L562).

It's not 30-45 seconds bad, but an initial load still takes ~3 seconds. Subsequent requests are faster, but pagespeed visits your site usually only once with an empty cache and not multiple times.
